File: src/main/java/com/mark/map/Region.java
package com.mark.map;

import com.mark.Buffer;

import java.util.Arrays;

public final class Region {


    public static int[] COSINE = new int[2048];


    private final int width;
    private final int length;
    private final int[][][] tileHeights;
    public final byte[][][] tileFlags;

    private final short[][][] underlays;
    private final short[][][] overlays;
    private final byte[][][] overlayShapes;
    private final byte[][][] overlayOrientations;
    public byte[][][] manualTileHeight;

    public Region(byte[][][] tileFlags, int[][][] tileHeights) {
        width = 104;
        length = 104;

        this.tileHeights = new int[4][width + 1][length + 1];
        this.tileFlags = new byte[4][width][length];
        underlays = new short[4][width][length];
        overlays = new short[4][width][length];
        manualTileHeight = new byte[4][width][length];
        overlayShapes = new byte[4][width][length];
        overlayOrientations = new byte[4][width][length];

    }




    public final void unpackTiles(byte[] data, int dX, int dY, int regionX, int regionY) {

        Buffer buffer = new Buffer(data);


        for (int z = 0; z < 4; z++) {
            for (int localX = 0; localX < 64; localX++) {
                for (int localY = 0; localY < 64; localY++) {
                    readTile(buffer, localX + dX, localY + dY, z, regionX, regionY, 0);
                }
            }
        }


        this.setHeights();// XXX Fix for ending of region sloping down

    }

    public void setHeights() {
        // TODO Find a better way to fix the sloping issue

        for(int z = 0;z<4;z++) {
            for(int y = 0;y<=length;y++) {
                tileHeights[z][width][y] = tileHeights[z][width - 1][y];
            }


            for(int x = 0;x<=width;x++) {
                tileHeights[z][x][length] = tileHeights[z][x][length - 1];
            }

        }

    }


    private void readTile(Buffer buffer, int x, int y, int z, int regionX, int regionY, int orientation) {// XXX

        if (x >= 0 && x < 64 && y >= 0 && y < 64) {
            tileFlags[z][x][y] = 0;
            try {
                do {
                    int type = buffer.readUByte();
                    if (type == 0) {
                        manualTileHeight[z][x][y] = 0;
                        if (z == 0) {
                            tileHeights[0][x][y] = -calculateHeight(0xe3b7b + x + regionX, 0x87cce + y + regionY) * 8;
                        } else {
                            tileHeights[z][x][y] = tileHeights[z - 1][x][y] - 240;
                        }

                        return;
                    } else if (type == 1) {
                        manualTileHeight[z][x][y] = 1;
                        int height = buffer.readUByte();
                        if (height == 1) {
                            height = 0;
                        }
                        if (z == 0) {
                            tileHeights[0][x][y] = -height * 8;
                        } else {
                            tileHeights[z][x][y] = tileHeights[z - 1][x][y] - height * 8;
                        }

                        return;
                    } else if (type <= 49) {
                        overlays[z][x][y] = (short) buffer.readUByte();
                        overlayShapes[z][x][y] = (byte) ((type - 2) / 4);
                        overlayOrientations[z][x][y] = (byte) (type - 2 + orientation & 3);
                    } else if (type <= 81) {
                        tileFlags[z][x][y] = (byte) (type - 49);
                    } else {
                        underlays[z][x][y] = (short) (type - 81);
                    }
                } while (true);
            }catch(Exception e) {}
        }

        try {
            do {
                int in = buffer.readUByte();
                if (in == 0) {
                    break;
                } else if (in == 1) {
                    buffer.readUByte();
                    return;
                } else if (in <= 49) {
                    buffer.readUShort();
                }
            } while (true);
        }catch(Exception e) {}

    }

    public byte[] save_terrain_block() {
        Buffer buffer = new Buffer(new byte[131072]);
        for (int tile_y = 0; tile_y < 4; tile_y++) {
            for (int tile_x = 0; tile_x < 64; tile_x++) {
                for (int tile_z = 0; tile_z < 64; tile_z++) {
                    save_terrain_tile(tile_y, tile_x, tile_z, buffer);
                }

            }

        }

        byte[] data = Arrays.copyOf(buffer.getPayload(), buffer.getPosition());
        return data;
    }

    private static int calculateHeight(int x, int y) {
        int height = interpolatedNoise(x + 45365, y + 0x16713, 4) - 128
                + (interpolatedNoise(x + 10294, y + 37821, 2) - 128 >> 1) + (interpolatedNoise(x, y, 1) - 128 >> 2);
        height = (int) (height * 0.3D) + 35;

        if (height < 10) {
            height = 10;
        } else if (height > 60) {
            height = 60;
        }

        return height;
    }

    private void save_terrain_tile(int y, int x, int z, Buffer buffer) {
        if (overlays[y][x][z] != 0) {
            buffer.writeShort(overlayShapes[y][x][z] * 4 + (overlayOrientations[y][x][z] & 3) + 2);
            buffer.writeShort(overlays[y][x][z]);
        }
        if (tileFlags[y][x][z] != 0) {
            buffer.writeShort(tileFlags[y][x][z] + 49);
        }
        if (underlays[y][x][z] != 0) {
            buffer.writeShort(underlays[y][x][z] + 81);
        }
        if (manualTileHeight[y][x][z] == 1 || y == 0) {
            buffer.writeShort(1);
            if (y == 0) {
                buffer.writeByte(-tileHeights[y][x][z] / 8);
            } else {
                buffer.writeByte(-(tileHeights[y][x][z] - tileHeights[y - 1][x][z]) / 8);
            }
        } else {
            buffer.writeShort(0);
        }
    }


    private static int calculateVertexHeight(int i, int j) {
        int mapHeight = (interpolatedNoise(i + 45365, j + 0x16713, 4) - 128) + (interpolatedNoise(i + 10294, j + 37821, 2) - 128 >> 1) + (interpolatedNoise(i, j, 1) - 128 >> 2);
        mapHeight = (int) ((double) mapHeight * 0.29999999999999999D) + 35;
        if (mapHeight < 10) {
            mapHeight = 10;
        } else if (mapHeight > 60) {
            mapHeight = 60;
        }
        return mapHeight;
    }

    private static int interpolatedNoise(int x, int y, int frequencyReciprocal) {
        int l = x / frequencyReciprocal;
        int i1 = x & frequencyReciprocal - 1;
        int j1 = y / frequencyReciprocal;
        int k1 = y & frequencyReciprocal - 1;
        int l1 = smoothNoise(l, j1);
        int i2 = smoothNoise(l + 1, j1);
        int j2 = smoothNoise(l, j1 + 1);
        int k2 = smoothNoise(l + 1, j1 + 1);
        int l2 = interpolate(l1, i2, i1, frequencyReciprocal);
        int i3 = interpolate(j2, k2, i1, frequencyReciprocal);
        return interpolate(l2, i3, k1, frequencyReciprocal);
    }

    private static int interpolate(int a, int b, int angle, int frequencyReciprocal) {
        int cosine = 0x10000 - COSINE[(angle * 1024) / frequencyReciprocal] >> 1;
        return (a * (0x10000 - cosine) >> 16) + (b * cosine >> 16);
    }

    private static int smoothNoise(int x, int y) {
        int corners = calculateNoise(x - 1, y - 1) + calculateNoise(x + 1, y - 1) + calculateNoise(x - 1, y + 1) + calculateNoise(x + 1, y + 1);
        int sides = calculateNoise(x - 1, y) + calculateNoise(x + 1, y) + calculateNoise(x, y - 1) + calculateNoise(x, y + 1);
        int center = calculateNoise(x, y);
        return corners / 16 + sides / 8 + center / 4;
    }

    private static int calculateNoise(int x, int y) {
        int k = x + y * 57;
        k = k << 13 ^ k;
        int l = k * (k * k * 15731 + 0xc0ae5) + 0x5208dd0d & 0x7fffffff;
        return l >> 19 & 0xff;
    }


    public final void initiateVertexHeights(int yOffset, int yLength, int xLength, int xOffset) {
        for (int y = yOffset; y <= yOffset + yLength; y++) {
            for (int x = xOffset; x <= xOffset + xLength; x++) {
                if (x >= 0 && x < width && y >= 0 && y < length) {
                    if (x == xOffset && x > 0) {
                        tileHeights[0][x][y] = tileHeights[0][x - 1][y];
                    }
                    if (x == xOffset + xLength && x < width - 1) {
                        tileHeights[0][x][y] = tileHeights[0][x + 1][y];
                    }
                    if (y == yOffset && y > 0) {
                        tileHeights[0][x][y] = tileHeights[0][x][y - 1];
                    }
                    if (y == yOffset + yLength && y < length - 1) {
                        tileHeights[0][x][y] = tileHeights[0][x][y + 1];
                    }
                }
            }
        }
    }
}

File: src/main/kotlin/com/mark/map/Region.kt
package com.mark.map

import com.mark.utils.Buffer
import java.util.*

class Region {

    private val width = 104
    private val length = 104
    private val tileHeights: Array<Array<IntArray>> = Array(4) { Array(width + 1) { IntArray(length + 1) } }
    val tileFlags: Array<Array<ByteArray>> = Array(4) { Array(width) { ByteArray(length) } }
    private val underlays: Array<Array<ShortArray>> = Array(4) { Array(width) { ShortArray(length) } }
    private val overlays: Array<Array<ShortArray>> = Array(4) { Array(width) { ShortArray(length) } }
    private val overlayShapes: Array<Array<ByteArray>> = Array(4) { Array(width) { ByteArray(length) } }
    private val overlayOrientations: Array<Array<ByteArray>> = Array(4) { Array(width) { ByteArray(length) } }
    private var manualTileHeight: Array<Array<ByteArray>> = Array(4) { Array(width) { ByteArray(length) } }

    fun unpackTiles(data: ByteArray, dX: Int, dY: Int, regionX: Int, regionY: Int) {
        val buffer = Buffer(data)
        for (plane in 0..3) {
            for (x in 0..63) {
                for (y in 0..63) {
                    readTile(buffer, x + dX, y + dY, plane, regionX, regionY)
                }
            }
        }
        setHeights()
    }

    private fun setHeights() {
        for (plane in 0..3) {
            for (y in 0..length) {
                tileHeights[plane][width][y] = tileHeights[plane][width - 1][y]
            }
            for (x in 0..width) {
                tileHeights[plane][x][length] = tileHeights[plane][x][length - 1]
            }
        }
    }

    private fun readTile(buffer: Buffer, x: Int, y: Int, z: Int, regionX: Int, regionY: Int) {
        if (x in 0..63 && y in 0..63) {
            tileFlags[z][x][y] = 0
            do {
                val type = buffer.readUByte()
                if (type == 0) {
                    manualTileHeight[z][x][y] = 0
                    if (z == 0) {
                        tileHeights[0][x][y] = -calculateHeight(0xe3b7b + x + regionX, 0x87cce + y + regionY) * 8
                    } else {
                        tileHeights[z][x][y] = tileHeights[z - 1][x][y] - 240
                    }
                    return
                } else if (type == 1) {
                    manualTileHeight[z][x][y] = 1
                    var height = buffer.readUByte()
                    if (height == 1) {
                        height = 0
                    }
                    if (z == 0) {
                        tileHeights[0][x][y] = -height * 8
                    } else {
                        tileHeights[z][x][y] = tileHeights[z - 1][x][y] - height * 8
                    }
                    return
                } else if (type <= 49) {
                    overlays[z][x][y] = buffer.readUByte().toShort()
                    overlayShapes[z][x][y] = ((type - 2) / 4).toByte()
                    overlayOrientations[z][x][y] = (type - 2 + 0 and 3).toByte()
                } else if (type <= 81) {
                    tileFlags[z][x][y] = (type - 49).toByte()
                } else {
                    underlays[z][x][y] = (type - 81).toShort()
                }
            } while (true)
        }
        do {
            val opcode1 = buffer.readUByte()
            if (opcode1 == 0) {
                break
            } else if (opcode1 == 1) {
                buffer.readUByte()
                return
            } else if (opcode1 <= 49) {
                buffer.readUShort()
            }
        } while (true)
    }

    fun saveTerrainBlock(): ByteArray {
        val buffer = Buffer(ByteArray(131072))
        for (plane in 0..3) {
            for (x in 0..63) {
                for (y in 0..63) {
                    saveTerrainTile(plane, x, y, buffer)
                }
            }
        }
        return buffer.payload.copyOf(buffer.position)
    }

    private fun saveTerrainTile(y: Int, x: Int, z: Int, buffer: Buffer) {
        if (overlays[y][x][z].toInt() != 0) {
            buffer.writeShort(overlayShapes[y][x][z] * 4 + (overlayOrientations[y][x][z].toInt() and 3) + 2)
            buffer.writeShort(overlays[y][x][z].toInt())
        }
        if (tileFlags[y][x][z].toInt() != 0) {
            buffer.writeShort(tileFlags[y][x][z] + 49)
        }
        if (underlays[y][x][z].toInt() != 0) {
            buffer.writeShort(underlays[y][x][z] + 81)
        }
        if (manualTileHeight[y][x][z].toInt() == 1 || y == 0) {
            buffer.writeShort(1)
            if (y == 0) {
                buffer.writeByte(-tileHeights[y][x][z] / 8)
            } else {
                buffer.writeByte(-(tileHeights[y][x][z] - tileHeights[y - 1][x][z]) / 8)
            }
        } else {
            buffer.writeShort(0)
        }
    }

    companion object {

        private var COSINE = IntArray(2048)
        private fun calculateHeight(x: Int, y: Int): Int {
            var height = interpolatedNoise(x + 45365, y + 0x16713, 4) - 128 + (interpolatedNoise(
                x + 10294,
                y + 37821,
                2
            ) - 128 shr 1) + (interpolatedNoise(x, y, 1) - 128 shr 2)
            height = (height * 0.3).toInt() + 35
            if (height < 10) {
                height = 10
            } else if (height > 60) {
                height = 60
            }
            return height
        }


        private fun interpolatedNoise(x: Int, y: Int, frequencyReciprocal: Int): Int {
            val l = x / frequencyReciprocal
            val i1 = x and frequencyReciprocal - 1
            val j1 = y / frequencyReciprocal
            val k1 = y and frequencyReciprocal - 1
            val l1 = smoothNoise(l, j1)
            val i2 = smoothNoise(l + 1, j1)
            val j2 = smoothNoise(l, j1 + 1)
            val k2 = smoothNoise(l + 1, j1 + 1)
            val l2 = interpolate(l1, i2, i1, frequencyReciprocal)
            val i3 = interpolate(j2, k2, i1, frequencyReciprocal)
            return interpolate(l2, i3, k1, frequencyReciprocal)
        }

        private fun interpolate(a: Int, b: Int, angle: Int, frequencyReciprocal: Int): Int {
            val cosine = 0x10000 - COSINE[angle * 1024 / frequencyReciprocal] shr 1
            return (a * (0x10000 - cosine) shr 16) + (b * cosine shr 16)
        }

        private fun smoothNoise(x: Int, y: Int): Int {
            val corners = calculateNoise(x - 1, y - 1) + calculateNoise(x + 1, y - 1) + calculateNoise(
                x - 1,
                y + 1
            ) + calculateNoise(x + 1, y + 1)
            val sides = calculateNoise(x - 1, y) + calculateNoise(x + 1, y) + calculateNoise(x, y - 1) + calculateNoise(
                x,
                y + 1
            )
            val center = calculateNoise(x, y)
            return corners / 16 + sides / 8 + center / 4
        }

        private fun calculateNoise(x: Int, y: Int): Int {
            var k = x + y * 57
            k = k shl 13 xor k
            val l = k * (k * k * 15731 + 0xc0ae5) + 0x5208dd0d and 0x7fffffff
            return l shr 19 and 0xff
        }
    }
}